### ISSUE TITLE
Refactor annotation specific code to separate module(s)

### ISSUE THREAD

**Issue by julesjacobsen** [monarch-initiative/phenol#233]
@pnrobinson I think it might be a good idea to pull out `org.monarchinitiative.phenol.annotations.hpo` and `org.monarchinitiative.phenol.io.annotations.hpo` into an HPO-specific module which imports the phenol-core and phenol-io modules. These are used for producing HPO release data and are a bit out of place in what ought to be a 'phenotype ontology' library for other applications to use.

Also why is the `org.monarchinitiative.phenol.formats` package in the phenol-core module? This seems to be used only by classes in the phenol-io module.

Perhaps the phenol-io module is misnamed. It ought to be called phenol-annotations and the actual io part (`OntologyLoader`, `org.monarchinitiative.phenol.io.utils` and the top level classes in `org.monarchinitiative.phenol.io.obo`) would be better placed in the phenol-core module. This way client code would only require the phenol-core module to be able to load and work with an ontology and all the ontology annotation-specific code is in a separate jar/ jars if it seems like the HPO/GO/MP code is better-off contained in completely separate modules.

**Comment by pnrobinson** [comment]
Hi Jules,
the original motivation for putting the io.annotations.hpo here is that we need the same Q/C code both for ingesting the annotations for say LIRICAL as for making the phenotype.hpoa file. 
I agree that there are a lot of things we could do to improve the overall architecture of phenol. 
It would be great to discuss this via zoom at some point. My current suggestion is to move to Java 11 modules and to put a lot of the many details into packages that have zero outside visibility. I would also suggest we move to JSON as the input format and drop OWL API, since that is not (easily) Java 11 compliant as far as I can see (and is too slow). 
I do not know if we can completely afford to do this with our internal code base, but it would be good to do so soon if we want to, i.e., before others start depending on us staying with Java 8 forever.

**Comment by julesjacobsen** [comment]
I'm not suggesting we move them to a completely different repository, just a separate module which would depend on phenol-core. e.g. LIRICAL would import phenol-core and phenol-annotations in order to ingest the phenotype.hpoa file. Exomiser might want to only use the phenol-core library in exomiser-core for doing profile matching, but would need the phenol-annotations in the exomiser-data-phenotype module to import the phenotype.hpoa into the database, but not need it at runtime.

I think we're going to have some difficulty removing the OWL API as its a dependency of ontolib and that might need some refactoring first before we can fully remove OWL API. 

Yes, let's schedule a call.

**Comment by pnrobinson** [comment]
@julesjacobsen I think your suggestion in the top comment is good and will implement. 
Java 13 was released last week, and we still cannot do Java 9 with phenol, which will require more work and will almost certainly not be possible for the 1.4.2 release.

**Comment by julesjacobsen** [comment]
It depends on what you mean by 'do'. You should be able to update to compile with and utilise Java 11 language features. The only thing you can't do yet is use the module system.

**Comment by pnrobinson** [comment]
@julesjacobsen  This issue has gotten too diffuse. I think that phenol should support annotations from various sources, but perhaps it would be good to make a module called hporelease that will take care of the tasks that are specific to making the phenotype.hpoa file and perhaps related files in the future.  Annotation logic that is useful to general applications could stay where it is, at least for now.

**Comment by julesjacobsen** [comment]
@pnrobinson, yes we've become a bit side-tracked here! If you're happy with what I suggested originally, I'll go ahead with the changes.

**Comment by pnrobinson** [comment]
@julesjacobsen  I think that this is partially implemented. But looking over the code, there are other similar issues of modules that are too entangled with each other. I am wondering if this would be a good use case for the new Java module system, although that would entail a lot of refactoring.

**Comment by julesjacobsen** [comment]
Which other classes/modules are still entangled? It look like I committed all my changes into the devlop branch, but I forget where I got to with things.

**Comment by pnrobinson** [comment]
I had to comment out two of the unit tests (this is another issue). The analysis module should not be imported by any of the other modules and can import the IO module. IntelliJ was telling me that there was a circular dependency -- not sure if this was the case. I think that we might want to refactor a few things. One of the classes was declared final (which is a good idea for enterprise software but not really necessary for bioinformatics -- we are all friends!). If we unfinalize it, then I can mock the class and allow testing.

**Comment by julesjacobsen** [comment]
Can you be more specific about the branch and classes you're talking about? On the development branch there shouldn't be any circular dependencies. How did IntelliJ warn you?

**Comment by pnrobinson** [comment]
This was the issue:
https://github.com/monarch-initiative/phenol/issues/247
The test class H2ScoreDistributionReaderTest is in the module phenol-io. However, it requires classes from the annotation module. If you uncomment the code, you see that IntelliJ wants to add a dependency of the io module on the annotation module, which I do not think we want.

**Comment by pnrobinson** [comment]
I think we can close this issue. The H2ScoreDistributionReaderTest issue is here: https://github.com/monarch-initiative/phenol/issues/247 but this is not a critical issue.
